pkg/cmd/grafana/apiserver/auth/tokenAuthenticator.go

```diff
@@ -18,54 +25,169 @@ const (
 	extraKeyGLSA        = "glsa"
 )
 
-func NewAccessTokenAuthenticator(config *authn.IDVerifierConfig) authenticator.RequestFunc {
+// TODO: There are equivalent structs in auth package but go.mod was giving me grief
+// Those structs are counterintuitively called SignIDTokenRequest and SignIDTokenResponse
+// Just because access token signing is implemented pretty similar to ID token signing
+type signAccessTokenRequest struct {
+}
+
+type accessTokenData struct {
+	Token string `json:"token"`
+}
+
+// TODO: this should likely be defined in auth package, was defined in a test as local
+type signAccessTokenResponse struct {
+	Status      string          `json:"status"`
+	AccessToken accessTokenData `json:"data,omitempty"`
+	Error       string          `json:"error,omitempty"`
+}
+
+var _ authenticator.Request = &AccessTokenAuthenticator{}
+
+type AccessTokenAuthenticator struct {
+	serviceBaseURL string
+	systemCAPToken string
+	validator      *TokenValidator
+}
+
+// TODO: add tests
+func newAccessTokenAuthenticator(options *options.AuthnOptions, client *http.Client) *AccessTokenAuthenticator {
 	verifier := authn.NewVerifier[CustomClaims](authn.IDVerifierConfig{
-		SigningKeysURL:   config.SigningKeysURL,
-		AllowedAudiences: config.AllowedAudiences,
+		SigningKeysURL:   options.IDVerifierConfig.SigningKeysURL,
+		AllowedAudiences: options.IDVerifierConfig.AllowedAudiences,
 	})
-	return getAccessTokenAuthenticatorFunc(&TokenValidator{verifier})
+
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	return &AccessTokenAuthenticator{
+		validator:      &TokenValidator{verifier},
+		serviceBaseURL: options.ServiceBaseURL,
+		systemCAPToken: options.SystemCAPToken,
+	}
 }
 
-func getAccessTokenAuthenticatorFunc(validator *TokenValidator) authenticator.RequestFunc {
-	return func(req *http.Request) (*authenticator.Response, bool, error) {
-		accessToken := req.Header.Get(headerKeyAccessToken)
+func newAuthAPIAuthorizedRequest(method, path string, token string, body []byte) (*http.Request, error) {
+	buffer := bytes.NewBuffer(nil)
+	if len(body) > 0 {
+		buffer.Write(body)
+	}
+
+	req, err := http.NewRequest(method, path, buffer)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+
+	return req, nil
+}
+
+func (auth *AccessTokenAuthenticator) signAccessToken(tokenValidationResult *authn.Claims[CustomClaims]) (string, error) {
+	body, err := json.Marshal(signAccessTokenRequest{})
+	if err != nil {
+		return "", err
+	}
+
+	path, err := url.JoinPath(auth.serviceBaseURL, "/v1/sign-access-token")
+	if err != nil {
+		return "", err
+	}
+	req, err := newAuthAPIAuthorizedRequest(http.MethodPost, path, auth.systemCAPToken, body)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("X-Org-ID", tokenValidationResult.Rest.OrgId)
+
+	audience := strings.Split(tokenValidationResult.Claims.Audience[0], ":")
+	req.Header.Add("X-Realms", fmt.Sprintf(`[{"type":"%s","identifier":"%s"}]`, audience[0], audience[1]))
```

Comment:
does the `tokenValidationResult` ensure that `len(tokenValidationResult.Claims.Audience) > 0` and `len(audience) > 1`? if not, i'd suggest adding some checks here to avoid panics

Reply:
It doesn't. Thanks for the catch. Will add.